### PR TITLE
Remove wrong space in regex for Keytrade Bank PDF Importer

### PR DESCRIPTION
**Pull request by dimitriwauters**
Keytrade Bank PDF Importer:
There was some space character that induced some regex to not be matched.

### CODE REVIEW

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/KeytradeBankPDFExtractor.java [hunk 1]
```diff
@@ -64,17 +64,19 @@ private void addBuySellTransaction()
...
+                    t.setAmount(asAmount(v.get("amount")));
+                    t.setCurrencyCode(v.get("currency"));
                     t.setSecurity(getOrCreateSecurity(v));
                 })
 
                 // Ausführungsdatum und -zeit : 15/03/2021 12:31:50 CET
-                // Ordre créé à : 10/02/2021 15:23:42 CET
+                // Ordre créé à: 10/02/2021 15:23:42 CET
                 .section("date", "time")
-                .match("^(Ausf.hrungsdatum und \\-zeit|Ordre cr.. .) : (?<date>[\\d]{2}\\/[\\d]{2}\\/[\\d]{4}) (?<time>[\\d]{2}:[\\d]{2}:[\\d]{2}) .*$")
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/KeytradeBankPDFExtractor.java:77]: Hello @dimitriwauters 
I'm afraid I don't understand your changes.
In French, punctuation marks are always followed by spaces. As it is to be seen in the test files e.g. [Kauf02.txt - L13](https://github.com/buchen/portfolio/blob/53a59729777988c9e388dd08c274aac96b62db42/name.abuchen.portfolio.tests/src/name/abuchen/portfolio/datatransfer/pdf/keytradebank/Kauf02.txt#L13) and [L14](https://github.com/buchen/portfolio/blob/53a59729777988c9e388dd08c274aac96b62db42/name.abuchen.portfolio.tests/src/name/abuchen/portfolio/datatransfer/pdf/keytradebank/Kauf02.txt#L14). 
Therefore your changes are not correct, unless there is a correct PDF debug for it, which is not available yet.
In French, the grammar is that special characters are preceded by a whitespace. 
As an example: 
```
Ordre créé à : 19/10/2020 11:53:02 CET
Type d' ordre : Limit (16 EUR)
```

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/KeytradeBankPDFExtractor.java [hunk 2]
```diff
@@ -111,8 +113,9 @@ private void addBuySellTransaction()
                 })
 
                 // Auftragstyp : Limit (16 EUR)
+                // Type d'ordre: Limit (46,3 EUR
                 .section("note").optional()
-                .match("^(Auftragstyp|Type d' ordre) : (?<note>.*)$")
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/KeytradeBankPDFExtractor.java:115]: The same applies here

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/KeytradeBankPDFExtractor.java [hunk 3]
```diff
@@ -64,17 +64,19 @@ private void addBuySellTransaction()
...
+                    t.setCurrencyCode(v.get("currency"));
                     t.setSecurity(getOrCreateSecurity(v));
                 })
 
                 // Ausführungsdatum und -zeit : 15/03/2021 12:31:50 CET
-                // Ordre créé à : 10/02/2021 15:23:42 CET
+                // Ordre créé à: 10/02/2021 15:23:42 CET
                 .section("date", "time")
-                .match("^(Ausf.hrungsdatum und \\-zeit|Ordre cr.. .) : (?<date>[\\d]{2}\\/[\\d]{2}\\/[\\d]{4}) (?<time>[\\d]{2}:[\\d]{2}:[\\d]{2}) .*$")
+                .match("^(Ausf.hrungsdatum und \\-zeit|Ordre cr.. .)\\s?: (?<date>[\\d]{2}\\/[\\d]{2}\\/[\\d]{4}) (?<time>[\\d]{2}:[\\d]{2}:[\\d]{2}) .*$")
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/KeytradeBankPDFExtractor.java:79]: Hello @dimitriwauters 
I'm afraid I don't understand your changes.
In French, punctuation marks are always followed by spaces. As it is to be seen in the test files e.g. [Kauf02.txt - L13](https://github.com/buchen/portfolio/blob/53a59729777988c9e388dd08c274aac96b62db42/name.abuchen.portfolio.tests/src/name/abuchen/portfolio/datatransfer/pdf/keytradebank/Kauf02.txt#L13) and [L14](https://github.com/buchen/portfolio/blob/53a59729777988c9e388dd08c274aac96b62db42/name.abuchen.portfolio.tests/src/name/abuchen/portfolio/datatransfer/pdf/keytradebank/Kauf02.txt#L14). 
Therefore your changes are not correct, unless there is a correct PDF debug for it, which is not available yet.
In French, the grammar is that special characters are preceded by a whitespace. 
As an example: 
```
Ordre créé à : 19/10/2020 11:53:02 CET
Type d' ordre : Limit (16 EUR)
```

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/KeytradeBankPDFExtractor.java [hunk 4]
```diff
@@ -111,8 +113,9 @@ private void addBuySellTransaction()
                 })
 
                 // Auftragstyp : Limit (16 EUR)
+                // Type d'ordre: Limit (46,3 EUR
                 .section("note").optional()
-                .match("^(Auftragstyp|Type d' ordre) : (?<note>.*)$")
+                .match("^(Auftragstyp|Type d' ordre)\\s?: (?<note>.*)$")
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/KeytradeBankPDFExtractor.java:118]: The same applies here

**Comment by dimitriwauters** [comment]
Hello @Nirus2000,
I've made theses changes because I wanted to import my PDF files into the application and I noticed that it does not work with Keytrade PDF as there is no space between the ":" and the text, as you can see there :
[Keytrade PDF](https://i.imgur.com/O595CeA.jpg)

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/KeytradeBankPDFExtractor.java [hunk 5]
```diff
@@ -64,17 +64,19 @@ private void addBuySellTransaction()
                 // Achat 310 HOME24 SE  INH O.N. (DE000A14KEB5) à 15,9324 EUR
                 // Vente 310 VERBIO VER.BIOENERGIE ON (DE000A0JL9W6) à 33,95 EUR
                 // Verkauf 22 SARTORIUS AG O.N. (DE0007165607) für 247 EUR
-                .section("isin", "name", "shares", "currency")
-                .match("^(Kauf|Achat|Verkauf|Vente) (?<shares>[\\.,\\d]+) (?<name>.*) \\((?<isin>[\\w]{12})\\) .* (?<currency>[\\w]{3})$")
+                .section("isin", "name", "shares", "amount", "currency")
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/KeytradeBankPDFExtractor.java:67]: Including the amount here is also wrong.
We generate by this part only the security in the appropriate currency, if this does not exist yet. 
The amount would then be the market value and would be overwritten further down in the [part](https://github.com/buchen/portfolio/blob/53a59729777988c9e388dd08c274aac96b62db42/name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/KeytradeBankPDFExtractor.java#L84).

To understand how the security is created or created in the corresponding currency in PP, you can take a look [here](https://github.com/buchen/portfolio/blob/53a59729777988c9e388dd08c274aac96b62db42/name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/AbstractPDFExtractor.java#L173).

**Comment by dimitriwauters** [comment]
They have made some changes recently on the UI, maybe they changed the PDF too.

**Comment by Nirus2000** [comment]
> Hello @Nirus2000, I've made theses changes because I wanted to import my PDF files into the application and I noticed that it does not work with Keytrade PDF as there is no space between the ":" and the text, as you can see there : [Keytrade PDF](https://i.imgur.com/O595CeA.jpg)

Okay... the create a PDF-Debug an build a testCase.
You can find the test cases [here](https://github.com/buchen/portfolio/blob/53a59729777988c9e388dd08c274aac96b62db42/name.abuchen.portfolio.tests/src/name/abuchen/portfolio/datatransfer/pdf/keytradebank/KeytradeBankPDFExtractorTest.java).

Video tutorial:
[Extract PDF documents for debugging](https://forum.portfolio-performance.info/t/buchungen-aus-pdf-dateien-importieren/38#tutorial-extrahieren)

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/KeytradeBankPDFExtractor.java [hunk 6]
```diff
@@ -64,17 +64,19 @@ private void addBuySellTransaction()
...
                 .assign((t, v) -> {
                     t.setShares(asShares(v.get("shares")));
+                    t.setAmount(asAmount(v.get("amount")));
+                    t.setCurrencyCode(v.get("currency"));
                     t.setSecurity(getOrCreateSecurity(v));
                 })
 
                 // Ausführungsdatum und -zeit : 15/03/2021 12:31:50 CET
-                // Ordre créé à : 10/02/2021 15:23:42 CET
+                // Ordre créé à: 10/02/2021 15:23:42 CET
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/KeytradeBankPDFExtractor.java:77]: Please do not line the comments, which will help in making the regex... You can simply insert a new one below it.

**Comment by dimitriwauters** [comment]
I've made the changes that you requested.
I've also added a new modification for the fee and tax because otherwise the price is wrong.

The sell/buy PDF works but not the dividende as there is no
`Compte-titres:4/260745 Titre:79137418/DE000A0JL9W6 Nr. Quit.: CPN / 583554`
and this match is marked as mandatory.

**Comment by Nirus2000** [comment]
Hello @dimitriwauters 
I think there are just too many changes in the source code that need to be made 
and it would not be productive to discuss them step by step through individual comments in your pull request. 
Please don't take this the wrong way, okay? :pray:
I hope you understand my arguments.

So it would be better if you close this pull request and post your 
PDF-debugs in the Portfolio-Performance-Forum -> [here](https://forum.portfolio-performance.info/t/pdf-import-von-keytrade-bank/15416) <-.
You can also add the PDF debug for dividend posting there.

I will do a separate pull request later. 

However I have a question about the sale pdf-debug.

If I see it correctly, the gross amount is less than the net amount.
Should not the credit be smaller?
```
Montant gross + 898,00 EUR
Crédit 908,64 EUR
```

Could it be that this is a tax refund and the other is a fee refund?
```
Taxe boursière - 0,35% - 3,14 EUR <-- tax refund?
Courtage - 7,50 EUR <-- fee refund?
```

Regards
Alex

**Comment by dimitriwauters** [comment]
Hello @Nirus2000,
Okay... I will do that when I have some time but I can't speak German so I will post them with English comments if that's fine with you.

`Should not the credit be smaller?`
Yes, indeed. I've made a mistake (sorry), it should be 887,36 EUR.
I'll post another PDF that I don't need to modify on the forum.

`Could it be that this is a tax refund and the other is a fee refund?`
There is always a minus in front on theses numbers, for buying and selling.
You can see it in the following screenshot: for [buying](https://i.imgur.com/VZGn797.png) and [selling](https://i.imgur.com/N4ftPW3.png).
It is not a refund, this is taken from the gross amount when selling and added when buying (as when buying the gross amount is negative).

I've rapidly looked into your changed and they look fine to me. The only issue is the credit amount, you can keep it that way (I saw that you modified the gross amount to correspond to the credit) or modify the credit for 887,36 EUR (sorry again for that).
You will see, the dividend is a little bit more complicated to implement, there is some changes ;)

**Pull request closed without merging**